### ISSUE TITLE
Sha 256 support

### ISSUE THREAD

**Issue by Byron** [Byron/gitoxide#281]
Even though the foundation is set, it needs another push to actually make it work with different kinds of hashes.

### Tasks

* [ ] remove hash-type specific methods from `git-hash` and replace them with parametric usage of `git_hash::Kind`
* [ ] all code assuming hashes of len 20 should receive this value as parameter instead. This is what git does for the old index and pack file formats.
* [ ] general tests for reading  refs and objects of different len
* [ ] tests for writing and reading objects of different len, maybe even write a conversion program

**Comment by shea256** [comment]
+1 to this feature. Excited to see this implemented.

**Comment by msrd0** [comment]
+1. Given that git no longer considers sha256 support experimental and forges such as forgejo added support, this has became more relevant for new git repositories.

**Comment by D4V1D3-008** [comment]
+1. Without sha256 support Starship prompt cannot detect repositories using that algorithm.

**Comment by jokeyrhyme** [comment]
@Byron 

> make sure once Sha256 is added as ObjectId variant, that it's behind a feature toggle to allow builds that opt-out of SHA256 support to not unnecessarily use more memory than needed. Maybe there are alternatives to this, too.

should the old SHA1 support also be put behind a feature toggle (enabled by default), so that builds can choose to _only_ support SHA256 ?

**Comment by Byron** [comment]
> +1. Without sha256 support Starship prompt cannot detect repositories using that algorithm.

Starship would definitely be a great motivation to finally go through with this.

`gix-hash` has been created with that in mind:

https://github.com/GitoxideLabs/gitoxide/blob/ec69c88fc119f3aa1967a7e7f5fca30e3ce97595/gix-hash/src/object_id.rs#L13-L14

Right now it's an enum and one could add another variant to it, behind a feature toggle, as the size of these hash digests definitely affects the resource requirements of operations like `clone`. As the size of these hashes would increase, this would affect the memory consumption even when cloning SHA1 repositories (unfortunately).
Maybe there are better ways to represent this - I think Git simply allocates for it which would be one solution here as well.

In any case, if this were to move forward, I think it would be good to review the hash representation first, as it's very relevant for performance.

> [@Byron](https://github.com/Byron)
> 
> > make sure once Sha256 is added as ObjectId variant, that it's behind a feature toggle to allow builds that opt-out of SHA256 support to not unnecessarily use more memory than needed. Maybe there are alternatives to this, too.
> 
> should the old SHA1 support also be put behind a feature toggle (enabled by default), so that builds can choose to _only_ support SHA256 ?

I see the point and `gitoxide` should allow to be tailored to what application developers need. However, it by default it will try to be compatible to Git which means SHA1 *and* SHA256 must be supported, which is certainly what this would start out as.
Once there is explicit demand, I think a feature toggle for turning off SHA1 support could be contributed. This will be some work though, as the test-suite absolutely refers to it in plenty of places.

**Comment by jokeyrhyme** [comment]
https://about.gitlab.com/blog/what-s-new-in-git-2-51-0/

> ## SHA-256 as the default hash function
>
> The Git version control system stores objects in a content-addressable filesystem. This means it uses the hash of an object to address content such as files, directories, and revisions, unlike traditional filesystems, which use sequential numbers. Using a hash function has the following advantages:
> 
> Easy integrity checks as a single bit flip would change the hash output completely.
> 
> Fast object lookup as objects can be indexed by their hash.
> 
> Object names can be signed and third parties can trust the hash to address the signed object and all objects it references.
> 
> Communication using Git protocol and out of band communication methods have a short reliable string that can be used to reliably address stored content.
> 
> Since its inception, Git has used the SHA-1 hashing algorithm. However, security researchers have discovered some flaws in SHA-1, specifically the [SHAttered attack](https://shattered.io/), which shows a practical SHA-1 hash collision. We moved to using a hardened SHA-1 implementation by default since Git 2.13.0. However, SHA-1 is still a weak hashing algorithm and it is only a matter of time before additional attacks will further reduce its security.
> 
> SHA-256 was identified as the successor to SHA-1 in late 2018. Git 2.51.0 marks it as the default hash algorithm to be used in Git 3.0.

**Comment by Byron** [comment]
Indeed, work should be underway to get reftable and SHA256 support.

**Comment by jokeyrhyme** [comment]
i did have a quick attempt at `enum`-ification of gix-hash to support SHA256 hashes, however, it immediately broke this test: https://github.com/GitoxideLabs/gitoxide/blob/c2c8c2fb13749b2f10d882cdbadafcb2e4c4dba3/gix/src/id.rs#L200

i effectively had

```rust
enum Hash {
    Sha1( /* https://crates.io/crates/sha1-checked */ ),
    Sha256 ( /* https://crates.io/crates/sha2 */ ),
}
```

which blows up the `std::mem::size_of()` result

are we okay with substantially increasing memory usage for all users even if they don't use SHA256 ?

**Comment by Byron** [comment]
> Indeed, work should be underway to get reftable and SHA256 support.

Actually, it turns out I misread, and this really is only coming with Git 3, so no haste :).

Regarding the data-structure, the SHA256 variant only consumes 33 bytes, 1 byte for the tag, so it's basically perfect.
We would implement it with a feature toggle so that people with only SHA1 repos don't need to put more memory into their computers for cloning.

The current version of `gix::ObjectId` costs exactly 20 bytes and is thus perfectly sized, with the version that supports both hashes will nearly double the memory needs (and who knows what happens when padding starts playing a role).

**Comment by Skgland** [comment]
> > Indeed, work should be underway to get reftable and SHA256 support.
> 
> Actually, it turns out I misread, and this really is only coming with Git 3, so no haste :).


Isn't that a bit of a chicken and egg problem as <https://git-scm.com/docs/BreakingChanges> states (regarding reftable support)(highlighted in bold by me)

> A prerequisite for this change is that the ecosystem is ready to support the "reftable" format. Most importantly, alternative implementations of Git like JGit, libgit2 and **Gitoxide** need to support it.

**Comment by Byron** [comment]
RefTable work will start this year.

**Comment by lorenzleutgeb** [comment]
> Once there is explicit demand, I think a feature toggle for turning off SHA1 support

TBH I think this might be the wrong way round. There should be a feature that *enables* SHA-1, which should be added to the list of *default features* at the same time. A feature flag that *removes* features violates the principle that Cargo features must be additive (see the section on [Feature Unification in the Cargo Book](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification)). The situation is analogous to `no_std` (bad) vs. `std` (good). This would mean a breaking change to dependents that purposely disable default features, but it would be very easy for them to recover, enabling the new SHA-1 feature flag.

In a new crate (for the curious: https://github.com/lorenzleutgeb/heartwood/commit/dcb67b0be37fbdd348916695a804c5e2060e7203), I require users to enable a `sha1` feature from the get go (and of course I add it to the default features of the crate) even though the crate itself is useless without it:
```rust
#[cfg(not(feature = "sha1"))]
compile_error!("The `sha1` feature is required. In the future, it might be possible to build this crate without support for SHA-1.");
```
This will catch users that choose to disable default features and force them to re-add the feature flag to their explicit list if they want to do anything meaningful with the crate (such as, well, compile it).

Update: I filed #2213

**Comment by Byron** [comment]
> TBH I think this might be the wrong way round. There should be a feature that _enables_ SHA-1, which should be added to the list of _default features_ at the same time. A feature flag that _removes_ features violates the principle that Cargo features must be additive (see the section on [Feature Unification in the Cargo Book](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification)). The situation is analogous to `no_std` (bad) vs. `std` (good). This would mean a breaking change to dependents that purposely disable default features, but it would be very easy for them to recover, enabling the new SHA-1 feature flag.

You are right, agreed.

**Comment by jackieismpc** [comment]
I see that the current C version of git uses a global pointer variable as the current hash algorithm (pointing to sha1 or sha256). Is the approach we are introducing also like this?

**Comment by Byron** [comment]
No - C-Git can assume it's a CLI and thus uses globals for its own convenience, but that doesn't work in library crates. So everything is passed down as parameters where needed.
(And even if it would work in library crates, we'd never expose globals/`static` here, and even if it was a good idea it naturally breaks down when testing which makes it untenable).

**Comment by jackieismpc** [comment]
I understand your point, but using enums or similar wrappers and then explicitly passing parameters where needed might involve changes to the API in many places. Wouldn't such modifications be too drastic? Could you avoid this by specifying the upper layer and using a repository structure to construct the context for the intermediate operations? Also, I'd like to know how you're currently preparing to implement the conversion between SHA1 and SHA256 repositories. For example, the remote repository is SHA1, and what I pull to my local machine needs to be converted to SHA256.

**Comment by Byron** [comment]
> I understand your point, but using enums or similar wrappers and then explicitly passing parameters where needed might involve changes to the API in many places. Wouldn't such modifications be too drastic?

Maybe that would require drastic changes, after all, such a parameter would have to be passed to all places that need it. It's good that this was done from the very beginning so the codebase is prepared.

> Also, I'd like to know how you're currently preparing to implement the conversion between SHA1 and SHA256 repositories.

This is a separate issue and no work was done in this direction. It might be intrusive and some places in the depth of the codebase might have to change to perform these mappings, but… what needs to happen will happen.

**Comment by EliahKagan** [comment]
> C-Git can assume it's a CLI and thus uses globals for its own convenience, but that doesn't work in library crates.

In addition, I believe Git is trying to eliminate this kind of global state. Unfortunately I haven't been following this closely, but there are a number of patches [that remove uses](https://lore.kernel.org/git/?q=remove+the_repository) of the `the_repository` global variable and otherwise to [reduce Git's global state](https://lore.kernel.org/git/?q=%22Refactor+in+order+to+reduce+Git%27s+global+state%22).

My guess is that some architectural differences shall always remain between Git and gitoxide library crates in this area. For example, the `-C` option in `git` actually changes directory to the specified location. This has no good generallly safe analogue for a library (that runs in the same process as the caller). But in areas where the functionality is analogous, I suspect that Git will continue to gradually reduce the use of global state.

**Comment by Byron** [comment]
*(in response to a question on Zulip which I wanted to have here for posterity)*

Yes, I think so, too, removing `sha1` from the default features would be needed here.

But I don't like it too much because it would mean that each consumer of anything that depends on `gix-hash` now has to make a choice, by pulling in `gix-hash` and select the hashing function.
Alternatively, and probably more proper, each consumer of `gix-hash` has to duplicate the `sha1` and `sha256` features, pass them on, *and* fail at compile time if nothing is set as none of these can make a default selection.

Either way, for `[dev-dependencies]` it's trivial to to pull in 'themselves' as dependency and set the features they need, or pull in `gix-hash` and set it up accordingly, depending on whether `sha1` and `sha256` should be duplicated downstream, which I think would be the proper way.

So ultimately, it's very much a breaking change, and every user of anything that exposes or uses hashes needs to select the hashes they are interested in. `gix` can easily define `sha1` as default though, just plumbing cannot do that.

However… since the value of this really is *only* to be able to make pure `sha256` only builds, I think it's better to do this (i.e. make `sha1` and `sha256` available in each consumer) only for one crate at a time, smallest to biggest, and *also* add the SHA256 tests.

The first crate to get that treatment would also have to create utilities to run the same tests twice, once for each hash kind.

**Comment by A248** [comment]
I'm not sure why there is so much consternation about feature selection and micro-optimizing memory usage. It seems fairly overblown and unnecessary, and most library consumers should be plenty satisfied with dual sha1/sha256 support (presumably, the default features would enable both).

The situation you are describing where features proliferate to consumers just doesn't seem very relevant. If they don't care, they don't care and just use default features. If they _do_ care -- because they have specific requirements, need ultra-maximum performance, and only 1 compiled algorithm -- then they should be more than happy to enable a couple feature flags in Cargo.toml. It's a small price to pay if their niche use case is demanding it.

I'm talking about external consumers there. Of course, I can understand if gitoxide's own crates would be annoying to add feature flags for, since this repository is divided into dozens of crates. This ultimately comes down to maintenance decisions and the worth of splitting up the code so profusely. Features introduce yet more "code splitting," and maybe you should weigh the costs of further splitting versus the maintenance burden of organizing so many dependencies and their features.

I mean, if I were you, I wouldn't be introducing any features at all. I'd just bake in sha1 and sha256 into the crate itself. Much simpler, and if anyone complains, they can PR optimizations to the Rust compiler to eliminate [enum variants that are never constructed](https://www.reddit.com/r/rust/comments/o3igpb/do_enum_variants_get_optimized_away_if_never/). But you don't have to follow my advice.

**Comment by Byron** [comment]
Thanks for bringing this up.

It's fair to take a step back from this and ask what the value proposition is. Right now, the discriminator of the enum of `ObjectId` is one byte, and with sha256 enabled as well it (interestingly) becomes two bytes. I assume it would go back to one byte if sha1 was disabled. It really isn't a big deal.

Then there is dependencies, `sha1-checked` and `sha2`. Less is more there, but one can argue that supporting all Git hashes at all times is well worth it, and there is no need to support niche use-cases that probably are related to extreme niche optimisation. Others might say that they'd never want to support hashes that are known to be broken beyond a transition period, and not having support compiled in is one way to make very sure it's not going to be used.

> The situation you are describing where features proliferate to consumers just doesn't seem very relevant. If they don't care, they don't care and just use default features. If they _do_ care -- because they have specific requirements, need ultra-maximum performance, and only 1 compiled algorithm -- then they should be more than happy to enable a couple feature flags in Cargo.toml. It's a small price to pay if their niche use case is demanding it.

That's exactly the problem - once any crate in this repository requires a default feature to work, that default has to be enabled in the crate that is defining it - `gix-hash` in this case. This also makes it impossible to turn that feature off in that dependency graph - one can only ever enable them.

To make these features useful, they can't really be default, or else one might as well not have the features at all which is what you are proposing. Of course, I really like that for its simplicity.

But as `gitoxide` is a library, I also really like that people can chose to build binaries without support for outdated hashes. This could also be accomplished by negative Cargo features, like `no-sha1`, and since these would then also be additive, I think it would be fair game.

Doing this also means we'd instantly ship the `sha2` dependency, unless `no-sha256` is enabled. But most importantly, these negative features could be added in future once there is actual demand and maybe better compiler or Cargo support to do such a thing. Maybe in future, Cargo can even allow those who compile the binary disable features for the whole dependency tree.

Circling back to additive `sha1` and `sha256` features, I think the main act would be staged in Cargo manifests, as most code isn't actually interested in their specific size. So conditional code is going to be rare.

And this could be the argument that sways me towards staying on course, and to wire up optional features that are repated in each downstream `gix-` crate, so the last one in the chain can make the choice. Then `gix` (the crate) ships with defaults as always, but that's where users can later make their choice.

The cost of this is probably going to be quite low, it's just… work.

CC @cruessler who should really have a say in this. Negative cargo features? Would that be better fitting for the usecase? It really does feel like it, but that would make it impossible to also shed the respective dependencies. So there really is no other choice but to use cargo features, or to not use them at all.

**Comment by jokeyrhyme** [comment]
should it actually be possible at all to build gitoxide without SHA256 support? i.e. should functionality that is core to the way `git` works be able to be turned on and off?

or, put another way, if feature flags and how to use them here is such a complicated matter, should the work here be delayed or made more complex due to the use of feature flags?

**Comment by Byron** [comment]
Thanks for sharing!

After all the text [here](https://github.com/GitoxideLabs/gitoxide/issues/281#issuecomment-3713172561) I made myself believe that using feature toggles actually isn't much of an overhead beyond the `gix-hash` crate, as the types like `gix_hash::Kind` abstract over the actual hash, so typical parameters of interest like its size can easily be queried.

All consumers of `gix-hash` merely have to assume that the types they need exist, that's all, their code doesn't change. The tests are then going to clearly indicate, via `[dev-dependencies]`, which hashes are tested, as they will have to enable a hash or it won't compile.

That aspect, that consumers of the crate now have to chose a hash, is the major hassle for the downstream that does use plumbing directly. That's the fewest of power users, and they can probably cope as the first error they see is a compile error that clearly states what they have to do to fix it.

And that, I think, is a good thing as such plumbing code might be dependent (unknowingly) on the length of the hash, maybe they have hardcoded it even, and that means they have to think about the upgrade, instead of just accidentally failing to handle sha256.

And that… is the final argument I need to firmly believe feature toggles for hashes without defaults within `gix-plumbing` crates is the way to go.

Finally, let me note that the `gix` crate won't break for users, it will auto-select a hash, or hashes once these are known to be fully implemented, and thus always support everything Git supports unless users disable default features to make their own choices.

**Comment by cruessler** [comment]
Just wanted to leave my thoughts, though I don’t have to add that much:

- From what I can see, most of the implementation work might have been in `gix-hash`, and that has already been mostly done (expect for the removal of `sha1` from default features).
- Writing tests for both algorithms in downstream crates, from what I can see, represents the bulk of the work that still needs to be done.
- There might be cases where people really only want one of the available algorithms, and that’s possible with feature flags as far as I’m aware.
- I feel like negative cargo features would come with their own problems, at least conceptually. We would either have `no-sha1`, but `sha256` which wouldn’t be consistent. Or we would have `no-sha1` and `no-sha256` which doesn’t really make sense at this point as we don’t even have SHA256 support yet.

Of course there might be downsides to using feature flags the way we started implementing them that I don’t have on my radar yet, but given the above, I’m in favor of sticking to the initial plan.

**Comment by Byron** [comment]
@cruessler I agree with everything you said and think we are on the same page.

> I feel like negative cargo features would come with their own problems, at least conceptually. We would either have no-sha1, but sha256 which wouldn’t be consistent. Or we would have no-sha1 and no-sha256 which doesn’t really make sense at this point as we don’t even have SHA256 support yet.

Indeed, these can't be done as features really have to be additive. Everything else is surprising. The killer for the 'no-sha1` idea though is that it can't remove dependencies from the tree, so it would be pulled in and compiled (albeit not used or linked into the final binary). So I think biting the bullet for plumbing crates is the way to go.

**Comment by bk2204** [comment]
I'll just mention that Git 3.0, which is likely to arrive around September 2026, will default to creating SHA-256 repositories (and reftable).  Cloning existing SHA-1 repositories will clone them as SHA-1 repositories (since the interoperability code in upstream Git is not complete yet), but they will be created with reftable and all repositories created with `git init` will default to SHA-256 and reftable.

I know work is underway, but I just don't want to leave anyone off guard about the expected timeline here, since it's less than six months away.

**Comment by cruessler** [comment]
Thanks for the heads-up! I’ll try to make sure that SHA-256 support stays high on my list of priorities.

**Comment by Byron** [comment]
The plan is now here: https://github.com/GitoxideLabs/gitoxide/blob/main/etc/plan/sha256-support.md, superseding this issue.

**Comment by yonas** [comment]
@Byron Why not use a tracking issue instead of a doc?

**Comment by EliahKagan** [comment]
@Byron I think it's misleading and carries various disadvantages to have this issue and #2351 be closed as "not planned," when they are still planned *and* indeed have substantial progress already, just the itemized plan progress notes are being maintained in files in the repo. I do acknowledge that there may be reasons for this approach that I haven't considered, but I strongly suspect that there is a reasonably simple way to achieve the goal while keeping the issues open.

I'm guessing you closed these as "not planned" to avoid giving the impression that the checkbox-state of the plan items in these issues reflects or will necessarily continue to reflect the current progress. Assuming that's the motivation, and that the progress is going to be tracked primarily in those two docs rather than these two issues, I think a better approach may be to edit the issue descriptions to link to the docs, collapse the original plan under `<details>`, change the `[ ]` and `[x]` in the list to emojis or even just remove them (if collapsing under `<details>` is not enough), and/or even remove the list from the issue descriptions entirely.

People get a lot of notifications, and the probable result for many people may be to see that the issue is marked a "not planned" and assume that the functionality is no longer planned. Having the issue closed also makes it unclear whether this remains an appropriate place to comment on its progress, and I think there's a considerable disadvantage to missing out on comments such as https://github.com/GitoxideLabs/gitoxide/issues/281#issuecomment-4174048688. Not having open issues for major ongoing pushes may also make it harder to make clear to everyone when they *are* completed.

(While not core to the reasons I'd prefer this remain open, it's worth mentioning that I've also linked to this issue--including in a recent presentation--to show gitoxide's commitment to the security and compatibility benefits of supporting SHA256 OIDs, and to encourage Git-related software that has not yet begun to support them to consider doing so.)

**Comment by Byron** [comment]
You are right, sorry for the hassle. I was in "a mode" that prevented me to see viable alternatives.
It's back, and it will instead be synced to take the contents of the file occasionally.

**Comment by lunagl** [comment]
Hey, I'm interested in contributing to this effort. Are there currently any todos that might be suitable for a new contributor to tackle?

**Comment by cruessler** [comment]
@lunagl Thanks for your interest! Are you looking for particular areas to contribute to or specific kinds of tasks? I’m asking because my own approach so far has been to only plan one or two steps into the future, mostly deciding what to pick up next once I’ve finished a particular task. And that means that I don’t have a list of tasks ready in my head. :-) So I’d recommend you either have a look at the list in the first post to see whether any of the remaining tasks sound interesting to you, or you can describe the kinds of tasks you’re looking for, and then I can have a look at the list to see whether there might be anything that fits your description. What do you think?

**Comment by lunagl** [comment]
@cruessler thanks for the reply! I'm not entirely sure, I mostly would love to see sha256 fully supported in gitoxide and was wondering what open issues might be good to look into (to avoid starting on things that others are already doing or planning to do, or things where a lot of experience with the project would would be ideal, etc).

**Comment by cruessler** [comment]
@lunagl If this is about seeing SHA-256 support implemented as quickly as possible, then at this point it’s probably best to wait as the limiting factors are review capacity as well as people with deep knowledge of the codebase and its processes. It could potentially be useful, though, if you had a project that depends on `gitoxide`, to see whether the current level of SHA-256 support would already be sufficient for your use cases.

**Comment by 10ne1** [comment]
Hello, I posted https://github.com/GitoxideLabs/gitoxide/pull/2627 as my first contribution to fix the following item in the first post:

`Make SHA256-only builds compile where supported by feature flags.`

I plan to do some more work on the transport layer, to get commands like the follwing pass:

```
  � GIX_TEST_FIXTURE_HASH=sha256 cargo nextest run -p gix --features 
 async-network-client --no-fail-fast
  � GIX_TEST_FIXTURE_HASH=sha256 cargo nextest run -p gix --features 
 blocking-network-client --no-fail-fast
```

**Comment by 10ne1** [comment]
I'm mostly done with the protocol/transport parts, just pending review feedback.

Next I plan to work on the object database which fails under GIX_TEST_FIXTURE_HASH=sha256.

**Comment by Byron** [comment]
Moved to [this Feature Proposal Discussion](https://github.com/GitoxideLabs/gitoxide/discussions/2780). This issue is being closed while the proposal is reviewed; it will become an issue again once review is complete.

_AI disclosure: This comment was written and posted by OpenAI Codex on behalf of Byron._
